### ISSUE TITLE
Upgrade to 3.0 broke yadm installation

### ISSUE THREAD

**Issue by max-ae** [TheLocehiliosan/yadm#285]
### Describe the bug

I installed version 3.0 and followed the prompt to run `yadm upgrade` following the directory structure change.
This process errored and made yadm unusable. I cleared all files, reinstalled yadm and tried to clone my repository with `yadm clone`. I then get submodule related errors that reference directories from the old structure.

### To reproduce

1. Run `yadm upgrade` for a yadm repository with submodules
2. Upgrade fails
3. Clear old and new yadm directories
4. Try to clone repository again
5. See error

### Expected behavior

`yadm upgrade` not to break my installation and repository that worked perfectly fine with yadm 2.5.0

### Environment

 - Operating system: macOS 11.1
 - Version yadm: 3.0
 - Version Git: 2.30.0

### Additional context

Console Output: 
```
yadm clone git@github.com:max-ae/dotfiles.git
Leeres verteiltes Git-Repository in /Users/Max/.local/share/yadm/repo.git/ initialisiert
remote: Enumerating objects: 429, done.
remote: Counting objects: 100% (429/429), done.
remote: Compressing objects: 100% (252/252), done.
remote: Total 429 (delta 216), reused 336 (delta 123), pack-reused 0
Empfange Objekte: 100% (429/429), 687.19 KiB | 1.69 MiB/s, fertig.
Löse Unterschiede auf: 100% (216/216), fertig.
Von github.com:max-ae/dotfiles
 * [neuer Branch]    master     -> origin/master
fatal: Kein Git-Repository: .oh-my-zsh/custom/plugins/forgit/../../../../.config/yadm/repo.git/modules/.oh-my-zsh/custom/plugins/forgit
ERROR: Unable to fetch origin git@github.com:max-ae/dotfiles.git
```

**Comment by max-ae** [comment]
Restoring the old yadm repository from a backup and manually moving it to the new location without initialising or running other commands resolved this and made yadm work as before. 
However, this should not have happened in the first place (recommended upgrade method breaking installation).

**Comment by TheLocehiliosan** [comment]
@max-ae I'm sorry about this experience. I have done quite a bit of testing before this release, and I suspect this is something unique about you configuration. 

Can you help with a bit more information?

I think this is a repo requiring authentication, can you confirm your method of authentication?

What was the error during the upgrade? Is it a different error from the clone?

What are the locations of your submodules?

Is it possible to give me access to a repo that fails in this manor?

**Comment by max-ae** [comment]
Of course!

The repo is hosted on GitHub and uses SSH authentication. Cloning the repo via git as well as `yadm push` and `yadm pull` work just as expected (after manually restoring).

The error during the upgrade was something similar regarding submodules, but my Terminal scroll back was already cleared. After the upgrade, every yadm command errors with 
```
fatal: Kein Git-Repository: .oh-my-zsh/custom/plugins/forgit/../../../../.config/yadm/repo.git/modules/.oh-my-zsh/custom/plugins/forgit
```
as if something references the old directory. However, manually moving does not have this effect.
I read something on your migration website that all submodules get reinitialised, otherwise they would be broken, but that was not the case for me (I did not have to initialise them again after manually moving the directory).

The submodules lie in the following path: ~/.oh-my-zsh/custom/{plugins, themes}/, the error weirdly uses those paths to reference the yadm repo, I think.

I gave you access to https://github.com/max-ae/yadm-upgrade-error, which contains the exact submodule configuration that caused this error prior to manually migrating to the new directory structure.

**Comment by max-ae** [comment]
Another note: 
I noticed, that I in fact did not move the old directory, but rather copied it to the new location. yadm only works now if the repository is in both locations:
If I remove the old one, above error occurs again with every command (+ legacy path notice)
If I remove the new one, yadm complains about an uninitialised repo (+ legacy path notice)

**Comment by TheLocehiliosan** [comment]
@max-ae Thanks for providing this extra help for tracking down the problem. I do have access to the example repo you created.

I just did two initial tests... First I cloned your repo using https & yadm 3.0.0. There were no issues. I then initialized the submodules after cloning, and there was no issue.

Second, I cloned this repo using https & 2.5.0. I then initialized the submodules. Then I used 3.0.0 to run "yadm upgrade", and the repo was moved, and submodules relocated. No errors.

Were you able to test and experience the same errors using this new demo repo? I'll go back and setup SSH on my test image so I can test that way too (perhaps that matters).

**Comment by TheLocehiliosan** [comment]
@max-ae - I was able to setup SSH on my test image, and I was able to do both of those tests above using SSH instead. I haven't been able to re-create this problem using the test repo.

You mentioned you copied instead of moved. That is part of the reason the upgrade process re-initializes submodules. They cannot simply be moved. The yadm repo itself will have paths inside ~/.local/share/yadm/repo.git/config for each of the submodules.

You could remove those entries manually, and re-initialize the submodules. Or you could run `yadm submodule deinit -f $HOME` (which would also remove them, this is done within the upgrade process). And then reinitialize them.

I'm wondering if this has something to do with zsh configurations itself, I'll test next using zsh as my shell. I'm sure once I can get the error to happen, I'll be able to determine what the cause is and fix it.

**Comment by TheLocehiliosan** [comment]
@max-ae - I've been able to test while running zsh as my shell. I thought maybe there would be an issue with some files locked or something odd like that. But cloning with 3.0.0 or upgrading with 3.0.0 seem to work just fine. The zsh config is incomplete those, as there are some errors like `.oh-my-zsh/oh-my-zsh.sh` is missing (I think it's just not part of your demo repo).

I do notice that the error you initially posted concerns `.oh-my-zsh/custom/plugins/forgit`, but that is not part of this demo repo. Is that just an accidental omission?

**Comment by max-ae** [comment]
My zsh config uses [the Oh-my-zsh framework](https://github.com/ohmyzsh/ohmyzsh) which installs the `.oh-my-zsh` folder including `.oh-my-zsh/oh-my-zsh.sh`, therefore it's not managed by my yadm repo. Only the `.oh-my-zsh/custom/` folder which contains plugins not bundled with the main installer are managed by my yadm configuration.

And yes, I accidentally omitted forgit because I was too thorough scrubbing my repo from personal information... I will update the repo to include the erroneous folder.

I also had a look at your `upgrade` code and executed `yadm submodule deinit -f $HOME` after deleting the yadm repo from the old location:
```
fatal: Submodule work tree './' contains a .git directory (use 'rm -rf' if you really want to remove it including all of its history)
```

`yadm submodule deinit -f` yields 
```
fatal: Use '--all' if you really want to deinitialize all submodules
```

`yadm submodule deinit -f --all` then returns

```
fatal: Submodule work tree '.oh-my-zsh/custom/plugins/autoupdate' contains a .git directory (use 'rm -rf' if you really want to remove it including all of its history)
```

This last error is the same that occurred during `yadm upgrade` (for this or a different submodule) which I was not able to recover in my shell history. So it seems like something fails during the submodule deinitialisation.

Manually deinitialising the submodules with the exact paths works (like `yadm submodule deinit -f ~/.oh-my-zsh/custom/plugins/autoupdate`). After I reinitialised all of them using `yadm submodule update --init --recursive`, yadm works like on 2.5.0 without errors or legacy path notice.

So either there are some tweaks needed in your submodule reinitialisation code or something special in my setup prevented the existing code to work fine (though I do not know what that could be).

**Comment by max-ae** [comment]
Something I also noticed:
Your legacy path notice only talks about manually moving the files (as possibilities for migrating except running `yadm upgrade`). 
Maybe you should also include a hint about reinitialising submodules, if any exist, if one chooses to manually migrate.

**Comment by TheLocehiliosan** [comment]
Sure, that note about manually handling it could be good.

The `.git` within each submodule should actually be an ascii file which contains the path to the actual submodule repo under the modules directory in the yadm repo. The error that you are showing above indicates that the `.git` in your submodules is an actual directory. I can only create the same error you have above if I replace the `.git` file with a directory.

I'm re-reading through this, trying to understand the history of what has happened and how to prevent it for anyone else.

I think I've struck upon a scenario that seems to result in something close to what you experienced. These are the steps I took:

1. Cloned your test repo using 2.5.0
2. Initialized the submodules (again with 2.5.0, `yadm submodule update --init`)
3. I changed one of the submodules, removing it, and cloning it directly. This results in the `.git` directory containing the repo instead of it being a simple file, with a reference to the submodule location.
4. At this point `yadm status` shows that the submodule in question is modified
5. Then I added change `yadm add -u`, and committed them. This leaves stuff in a very strange state, that is, the `.git` for the submodule is an actual git directory instead of a reference.
6. `yadm upgrade` using 3.0.0 now failed when it got to that submodule.
7. `yadm clone` again using 3.0.0, now also fails, IF I DON'T REMOVE THE SUBMODULES FIRST

At this point, I strongly believe at least one of your submodule's `.git` file was replaced with an actual repo. I'm not sure what might have caused that. It it possible something else could recreate those repos (outside of initializing or adding the submodule with yadm)? I'm not familiar with this oh-my-zsh framework, does it manage and update those repos in some way?

**Comment by erijo** [comment]
How about running `git submodule absorbgitdirs` as a first step in the upgrade? Would that fix this issue?

**Comment by erijo** [comment]
I played around with the upgrade code a bit and changed it to try to deinit submodules before doing the move, and abort early if it fails. What do you think about doing something like that?

**Comment by max-ae** [comment]
First of all, sorry for taking so long to respond.

> At this point, I strongly believe at least one of your submodule's `.git` file was replaced with an actual repo. I'm not sure what might have caused that. It it possible something else could recreate those repos (outside of initializing or adding the submodule with yadm)? I'm not familiar with this oh-my-zsh framework, does it manage and update those repos in some way?

So basically, you install plugins for oh-my-zsh by cloning a repository inside the `~/.oh-my-zsh/custom/` directory. That means the repos are initialised by the user. They are then updated by the framework simply by performing `git pull` inside each plugin directory, no other operations.
As I am using yadm, I do not clone the repos directly, but rather add them as submodules to yadm. 
What I can imagine happened is that I installed some of the plugins before I was using yadm, therefore the directories had a `.git` directory, started to use yadm and added the directories as submodules, but git did not absorb the `.git` folders correctly, leading to the situation you described above, where a submodules `.git` is not a file.

**Comment by erijo** [comment]
The documentation for `git submodule absorbgitdirs` states:

> A repository that was cloned independently and later added as a submodule or old setups have the submodules git directory inside the submodule instead of embedded into the superprojects git directory.

So it seems like a case that can happen.

**Comment by erijo** [comment]
Furthermore, I think it would be good to not use `-f` to deinit unless the user has requested as it might throw away changes in the submodule. E.g. if oh-my-zsh has updated a plugin but this hasn't been committed with yadm, `deinit -f` followed by `init` will revert back to a previous commit. In this case it won't matter much as I guess oh-my-zsh will update the plugin again later, but it's easy to imagine cases where this would be very surprising (or worse) behavior.

**Comment by erijo** [comment]
I've managed to reproduce this with the following steps (using the docker image yadm/testbed:2020-12-29):
1. `yadm-2.5.0 init`
2. `git clone https://github.com/TheLocehiliosan/yadm.git a`
3. `yadm-2.5.0 submodule add https://github.com/TheLocehiliosan/yadm.git a`
4. `yadm-2.5.0 submodule add https://github.com/TheLocehiliosan/yadm.git b`
5. `yadm upgrade`

After this, `yadm status` reports `fatal: not a git repository: b/../.config/yadm/repo.git/modules/b`. Note that the naming is important. The cloned path must sort before the path used in step 4.

I've updated the test in #284 to reproduce this error. I've also updated my branch submodule-upgrade to include a call to `git submodule absorbgitdirs` which makes the test added in #284 pass.

**Comment by TheLocehiliosan** [comment]
@erijo I had not known about `submodule absorbegitdirs` before. This absolutely looks like the right approach. It seems like that feature was added in Git 2.12, and I know I have some people that still depend on Git 1.7. However, it looks like your changes will continue even if absorbegitdirs fails, so those using older versions of Git shouldn't be any worse off than they are today. I will try to get enough time today to merge these changes in.

**Comment by TheLocehiliosan** [comment]
I think these concerns have been addressed in the 3.0.2 release.

@max-ae - thanks for bringing attention to these issues, and for the additional testing and information.

@erijo - Thanks for identifying the absorbegitdirs solution.
